pkg/storage/couchdb/couchdbstore.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package couchdbstore
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	_ "github.com/go-kivik/couchdb" // The CouchDB driver
+	"github.com/go-kivik/kivik"
+
+	"github.com/trustbloc/edge-core/pkg/storage"
+)
+
+// Provider represents an CouchDB implementation of the storage.Provider interface
+type Provider struct {
+	hostURL       string
+	couchDBClient *kivik.Client
+}
+
+const blankHostErrMsg = "hostURL for new CouchDB provider can't be blank"
+
+// NewProvider instantiates Provider
+func NewProvider(hostURL string) (*Provider, error) {
+	if hostURL == "" {
+		return nil, errors.New(blankHostErrMsg)
+	}
+
+	client, err := kivik.New("couch", hostURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{hostURL: hostURL, couchDBClient: client}, nil
+}
+
+// OpenStore opens and returns a store for the given name.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	exists, err := p.couchDBClient.DBExists(context.Background(), name)
+	if err != nil {
+		return nil, err
+	}
+
+	if exists {
+		db := p.couchDBClient.DB(context.Background(), name)
+
+		if dbErr := db.Err(); dbErr != nil {
+			return nil, dbErr
+		}
+
+		return &CouchDBStore{db: db}, nil
+	}
+
+	err = p.couchDBClient.CreateDB(context.Background(), name)
+	if err != nil {
+		return nil, err
+	}
+
+	db := p.couchDBClient.DB(context.Background(), name)
+
+	if err := db.Err(); err != nil {
+		return nil, err
+	}
+
+	return &CouchDBStore{db: db}, nil
+}
+
+// CloseStore closes a previously opened store.
+// CouchDB and Kivik don't really have "close" mechanisms for individual databases, so there's nothing to do here.
+func (p *Provider) CloseStore(name string) error {
+	return nil
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.couchDBClient = nil
+	return nil
+}
+
+// StoreExists tells you whether a store with the given name is already open.
+func (p *Provider) StoreExists(name string) (bool, error) {
+	exists, err := p.couchDBClient.DBExists(context.Background(), name)
+	return exists, err
+}
+
+// CouchDBStore represents a CouchDB-backed database.
+type CouchDBStore struct {
+	db *kivik.DB
+}
+
+// Put stores the given key-value pair in the store.
+func (c *CouchDBStore) Put(k string, v []byte) error {
+	_, err := c.db.Put(context.Background(), k, v)
```

Comment:
you need to create doc and create key inside doc called data

Reply:
doc := map[string]interface{}{
        "_id": k,
        "data":v
    }

Reply:
@bstasyszyn @troyronda @llorllale @rolsonquadras

Reply:
[DB.Put()](https://godoc.org/github.com/go-kivik/kivik#DB.Put)

Reply:
Do we ever want to save the value as an attachment?

Reply:
https://github.com/trustbloc/fabric-peer-ext/blob/e486262f7eafc9247351e6f683885c627b829e80/pkg/collections/offledger/storeprovider/store/couchdbstore/dbstore.go#L249

Reply:
Per our discussion on Friday with @bstasyszyn , I updated the CouchDB store Put method so that a non-JSON payload will be saved as an attachment. I verified that JSON payloads do get merged with the CouchDB document (this should answer @bstasyszyn's question from Friday)